### PR TITLE
Add table azure_storage_container. closes #69

### LINKED ISSUE #69: Add table azure_storage_container
**References**
https://docs.microsoft.com/en-us/rest/api/storageservices/list-containers2

This will unblock the following CIS recommendations-
```
5.1.3 Ensure the storage container storing the activity logs is not publicly accessible (Automated)
5.1.4 Ensure the storage account containing the container with activity logs is encrypted with BYOK (Use Your Own Key) (Automated)

### PR DESCRIPTION
**Pull request by bigdatasourav**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
No env file present for the current environment:  staging 
 Falling back to .env config
No env file present for the current environment:  staging
customEnv TURBOT_TEST_EXPECTED_TIMEOUT 300

SETUP: tests/azure_storage_container []

PRETEST: tests/azure_storage_container

TEST: tests/azure_storage_container
Running terraform
data.azurerm_client_config.current: Refreshing state...
data.null_data_source.resource: Refreshing state...
azurerm_resource_group.named_test_resource: Creating...
azurerm_resource_group.named_test_resource: Creation complete after 2s [id=/subscriptions/3510ae4d-530b-497d-8f30-53b9616fc6c1/resourceGroups/turbottest76330]
azurerm_storage_account.named_test_resource: Creating...
azurerm_storage_account.named_test_resource: Still creating... [10s elapsed]
azurerm_storage_account.named_test_resource: Still creating... [20s elapsed]
azurerm_storage_account.named_test_resource: Still creating... [30s elapsed]
azurerm_storage_account.named_test_resource: Creation complete after 33s [id=/subscriptions/3510ae4d-530b-497d-8f30-53b9616fc6c1/resourceGroups/turbottest76330/providers/Microsoft.Storage/storageAccounts/turbottest76330]
azurerm_storage_container.named_test_resource: Creating...
azurerm_storage_container.named_test_resource: Creation complete after 0s [id=https://turbottest76330.blob.core.windows.net/turbottest76330]

Warning: Deprecated Resource

The null_data_source was historically used to construct intermediate values to
re-use elsewhere in configuration, the same can now be achieved using locals


Apply complete! Resources: 3 added, 0 changed, 0 destroyed.

Outputs:

resource_aka = azure:///subscriptions/3510ae4d-530b-497d-8f30-53b9616fc6c1/resourceGroups/turbottest76330/providers/Microsoft.Storage/storageAccounts/turbottest76330/blobServices/default/containers/turbottest76330
resource_aka_lower = azure:///subscriptions/3510ae4d-530b-497d-8f30-53b9616fc6c1/resourcegroups/turbottest76330/providers/microsoft.storage/storageaccounts/turbottest76330/blobservices/default/containers/turbottest76330
resource_id = /subscriptions/3510ae4d-530b-497d-8f30-53b9616fc6c1/resourceGroups/turbottest76330/providers/Microsoft.Storage/storageAccounts/turbottest76330/blobServices/default/containers/turbottest76330
resource_name = turbottest76330
subscription_id = 3510ae4d-530b-497d-8f30-53b9616fc6c1

Running SQL query: test-get-query.sql
[
  {
    "account_name": "turbottest76330",
    "deleted": false,
    "deny_encryption_scope_override": false,
    "has_immutability_policy": false,
    "has_legal_hold": false,
    "id": "/subscriptions/3510ae4d-530b-497d-8f30-53b9616fc6c1/resourceGroups/turbottest76330/providers/Microsoft.Storage/storageAccounts/turbottest76330/blobServices/default/containers/turbottest76330",
    "lease_duration": "",
    "lease_state": "Available",
    "lease_status": "Unlocked",
    "name": "turbottest76330",
    "public_access": "None",
    "remaining_retention_days": 0,
    "resource_group": "turbottest76330",
    "subscription_id": "3510ae4d-530b-497d-8f30-53b9616fc6c1",
    "type": "Microsoft.Storage/storageAccounts/blobServices/containers",
    "version": null
  }
]
✔ PASSED

Running SQL query: test-hydrate-query.sql
[
  {
    "id": "/subscriptions/3510ae4d-530b-497d-8f30-53b9616fc6c1/resourceGroups/turbottest76330/providers/Microsoft.Storage/storageAccounts/turbottest76330/blobServices/default/containers/turbottest76330",
    "name": "turbottest76330"
  }
]
✔ PASSED

Running SQL query: test-list-query.sql
[
  {
    "id": "/subscriptions/3510ae4d-530b-497d-8f30-53b9616fc6c1/resourceGroups/turbottest76330/providers/Microsoft.Storage/storageAccounts/turbottest76330/blobServices/default/containers/turbottest76330",
    "name": "turbottest76330"
  }
]
✔ PASSED

Running SQL query: test-not-found-query.sql
null
✔ PASSED

Running SQL query: test-turbot-query.sql
[
  {
    "akas": [
      "azure:///subscriptions/3510ae4d-530b-497d-8f30-53b9616fc6c1/resourceGroups/turbottest76330/providers/Microsoft.Storage/storageAccounts/turbottest76330/blobServices/default/containers/turbottest76330",
      "azure:///subscriptions/3510ae4d-530b-497d-8f30-53b9616fc6c1/resourcegroups/turbottest76330/providers/microsoft.storage/storageaccounts/turbottest76330/blobservices/default/containers/turbottest76330"
    ],
    "title": "turbottest76330"
  }
]
✔ PASSED

POSTTEST: tests/azure_storage_container

TEARDOWN: tests/azure_storage_container

SUMMARY:

1/1 passed.
```
</details>

# Example query results
<details>
  <summary>Results</summary>
  
```
> select name,id,type,account_name from azure_storage_container;
+---------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
| name                      | id                                                                                                                                                                            
+---------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
| insights-operational-logs | /subscriptions/3510ae4d-530b-497d-8f30-53b9616fc6c1/resourceGroups/turbot_rg/providers/Microsoft.Storage/storageAccounts/test1234qw/blobServices/default/containers/insights-o
| insights-activity-logs    | /subscriptions/3510ae4d-530b-497d-8f30-53b9616fc6c1/resourceGroups/turbot_rg/providers/Microsoft.Storage/storageAccounts/test1234qw/blobServices/default/containers/insights-a
+---------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
> 
> select
  name,
  id,
  type,
  account_name
from
  azure_storage_container
where
  public_access = 'None';
+------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------
| name | id                                                                                                                                                                      | type                     
+------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------
| test | /subscriptions/3510ae4d-530b-497d-8f30-53b9616fc6c1/resourceGroups/turbot_rg/providers/Microsoft.Storage/storageAccounts/test123aq/blobServices/default/containers/test | Microsoft.Storage/storage
+------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------

select
  name,
  id,
  type,
  account_name
from
  azure_storage_container
where
  has_legal_hold
+----------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| name           | id                                                                                                                                                                                      |
+----------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| test-container | /subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbot_rg/providers/Microsoft.Storage/storageAccounts/turbotrgdiag332/blobServices/default/containers/test-container |
+----------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```
> select
  name,
  id,
  type,
  account_name
from
  azure_storage_container
where
  lease_duration = 'Infinite';
+----------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| name           | id                                                                                                                                                                                      |
+----------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| test-container | /subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbot_rg/providers/Microsoft.Storage/storageAccounts/turbotrgdiag332/blobServices/default/containers/test-container |
+----------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

> select
  name,
  id,
  type,
  account_name
from
  azure_storage_container
where
  lease_state = 'Leased'
  or lease_state = 'Broken';
+----------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| name           | id                                                                                                                                                                                      |
+----------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| test-container | /subscriptions/d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8/resourceGroups/turbot_rg/providers/Microsoft.Storage/storageAccounts/turbotrgdiag332/blobServices/default/containers/test-container |
+----------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

</details>

### CODE REVIEW

File: azure/plugin.go [hunk 1]
```diff
@@ -61,6 +61,7 @@ func Plugin(ctx context.Context) *plugin.Plugin {
 			"azure_storage_table_service":       tableAzureStorageTableService(ctx),
 			"azure_subnet":                      tableAzureSubnet(ctx),
 			"azure_virtual_network":             tableAzureVirtualNetwork(ctx),
+			"azure_storage_container":           tableAzureStorageContainer(ctx),
```
Comment (Subhajit97) [inline azure/plugin.go:64]: Sort the list

File: azure/table_azure_storage_container.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+				Name:        "subscription_id",
+				Description: ColumnDescriptionSubscription,
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("ID").Transform(idToSubscriptionID),
+			},
+		},
+	}
+}
+
+//// FETCH FUNCTIONS
```
Comment (Subhajit97) [inline azure/table_azure_storage_container.go:98]:
```suggestion
//// LIST FUNCTION
```

File: azure/utils.go [hunk 3]
```diff
@@ -62,3 +62,9 @@ func toLower(_ context.Context, d *transform.TransformData) (interface{}, error)
 	valStr := types.SafeString(d.Value)
 	return strings.ToLower(valStr), nil
 }
+
+func idToAccountName(ctx context.Context, d *transform.TransformData) (interface{}, error) {
+	id := types.SafeString(d.Value)
+	account_name := strings.Split(id, "/")[8]
```
Comment (Subhajit97) [inline azure/utils.go:68]: Don't use snake case to define a variable, use camel case instead.
```suggestion
	accountName := strings.Split(id, "/")[8]
```

File: azure/table_azure_storage_container.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+			},
+			{
+				Name:        "akas",
+				Description: ColumnDescriptionAkas,
+				Type:        proto.ColumnType_JSON,
+				Transform:   transform.FromField("ID").Transform(idToAkas),
+			},
+
+			// Standard azure columns
+			{
```
Comment (Subhajit97) [inline azure/table_azure_storage_container.go:82]: Don't we get `location`?
If not then can we get the location from storage account?

File: azure/table_azure_storage_container.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+				Name:        "account_name",
+				Description: "The friendly name that identifies the storage account.",
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("ID").Transform(idToAccountName),
+			},
+			{
+				Name:        "container_properties",
+				Description: "The blob container properties.",
+				Type:        proto.ColumnType_JSON,
+				Transform:   transform.FromField("ContainerProperties"),
```
Comment (Subhajit97) [inline azure/table_azure_storage_container.go:58]: Can we split ContainerProperties object into multiple columns?

File: azure/table_azure_storage_container.go [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+			{
+				Name:        "name",
+				Description: "The friendly name that identifies the container.",
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "id",
+				Description: "The container ID",
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("ID"),
```
Comment (Subhajit97) [inline azure/table_azure_storage_container.go:40]:
```suggestion
				Transform:   transform.FromGo(),
```

File: azure-test/tests/azure_storage_container/variables.tf [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+variable "azure_environment" {
+  type        = string
+  default     = "public"
+  description = "Azure environment used for the test."
+}
+
+variable "azure_subscription" {
+  type        = string
+  default     = "d7245080-b4ae-4fe5-b6fa-2e71b3dae6c8"
```
Comment (Subhajit97) [inline azure-test/tests/azure_storage_container/variables.tf:16]:
```suggestion
  default     = "3510ae4d-530b-497d-8f30-53b9616fc6c1"
```

File: docs/tables/azure_storage_container.md [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+```sql
+select
+  name,
+  id,
+  type,
+  account_name
+from
+  azure_storage_container;
+```
```
Comment (Subhajit97) [inline docs/tables/azure_storage_container.md:17]:
```suggestion
```

```

File: docs/tables/azure_storage_container.md [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+### Ensure the storage container storing the activity logs is not publicly accessible
```
Comment (Subhajit97) [inline docs/tables/azure_storage_container.md:18]:
```suggestion

### Ensure the storage container storing the activity logs is not publicly accessible
```

**Review by Subhajit97 — CHANGES_REQUESTED** [review]
Make suggested changes

File: azure/table_azure_storage_container.go (hunk 4 above)
Comment (bigdatasourav) [reply azure/table_azure_storage_container.go:82]: as discussed will not have location field

File: azure/table_azure_storage_container.go (hunk 5 above)
Comment (bigdatasourav) [reply azure/table_azure_storage_container.go:58]: as discussed , split few columns which are important

File: azure/table_azure_storage_container.go [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+				Name:        "title",
+				Description: ColumnDescriptionTitle,
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("Name"),
+			},
+			{
+				Name:        "tags",
+				Description: ColumnDescriptionTags,
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("Etag"),
```
Comment (Subhajit97) [inline azure/table_azure_storage_container.go:102]: `Etag` is a system generated hash required for concurrency control, when we make an update or delete operation.
Remove `tags` column if it is not supported.

File: azure/utils.go [hunk 11]
```diff
@@ -62,3 +62,9 @@ func toLower(_ context.Context, d *transform.TransformData) (interface{}, error)
 	valStr := types.SafeString(d.Value)
 	return strings.ToLower(valStr), nil
 }
+
+func idToAccountName(ctx context.Context, d *transform.TransformData) (interface{}, error) {
```
Comment (Subhajit97) [inline azure/utils.go:66]: Are we using this function in any other tables?
If not, move this function to current table.

File: docs/tables/azure_storage_container.md [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+
+### Ensure the storage container storing the activity logs is not publicly accessible
+
+```sql
+select
+  jsonb_pretty(container_properties) as container_properties
+from
+  azure_storage_container
+where
+  name = 'insights-operational-logs';
```
Comment (Subhajit97) [inline docs/tables/azure_storage_container.md:27]: @bigdatasourav I think the output of the query is completely different from the heading that you have mentioned
It should be something like
```suggestion
select
  name
from
  azure_storage_container
where
  public_access = 'None';
```

File: azure/table_azure_storage_container.go (hunk 5 above)
Comment (Subhajit97) [reply azure/table_azure_storage_container.go:88]: Can you also move rest of the properties that are inside `ContainerProperties` to top level?
As `ContainerProperties` contains most of the container configurations, other than name and ID.

File: azure-test/tests/azure_storage_container/variables.tf [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+  storage_account_name  = azurerm_storage_account.named_test_resource.name
+  container_access_type = "private"
+}
+
+output "resource_aka" {
+  value = "azure:///subscriptions/${var.azure_subscription}/resourceGroups/${var.resource_name}/providers/Microsoft.Storage/storageAccounts/${var.resource_name}/blobServices/default/containers/${var.resource_name}"
+}
+
+output "resource_aka_lower" {
+  value = "azure:///subscriptions/${lower(var.azure_subscription)}/resourcegroups/${lower(var.resource_name)}/providers/microsoft.storage/storageaccounts/${lower(var.resource_name)}/blobservices/default/containers/${lower(var.resource_name)}"
```
Comment (Subhajit97) [inline azure-test/tests/azure_storage_container/variables.tf:63]:
```suggestion
  value = "azure:///subscriptions/${var.azure_subscription}/resourcegroups/${var.resource_name}/providers/microsoft.storage/storageaccounts/${var.resource_name}/blobservices/default/containers/${var.resource_name}"
```

**Review by Subhajit97 — CHANGES_REQUESTED** [review]
@bigdatasourav There are few more changes required. Please see comments, thanks!

File: azure/table_azure_storage_container.go (hunk 10 above)
Comment (bigdatasourav) [reply azure/table_azure_storage_container.go:102]: removed

**Review by Subhajit97 — APPROVED** [review]
LGTM

File: docs/tables/azure_storage_container.md [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+select
+  name,
+  id,
+  type,
+  account_name
+from
+  azure_storage_container
+where
+  public_access = 'None';
+```
```
Comment (rajmohanty17) [inline docs/tables/azure_storage_container.md:31]: Add more example 

1. List the containers which are either leased or lease broken state. If leased pls check if the lease_duration is infinite.
2. Check the retention days ending in next 7 days
3. List for which the legal_hold is true.

**Review by rajmohanty17 — CHANGES_REQUESTED** [review]
Pls implement the changes 
API link to test

File: docs/tables/azure_storage_container.md (hunk 14 above)
Comment (bigdatasourav) [reply docs/tables/azure_storage_container.md:31]: doc updated

**Review by rajmohanty17 — APPROVED** [review]
LGTM

**Review by cbruno10 — APPROVED** [review]

**Pull request merged**